File: arena-xunit/src/Lifecycle/ArenaLifecycleError.cs
using ArenaDotnet.Xunit.Ffi;

namespace ArenaDotnet.Xunit.Lifecycle;

public class ArenaLifecycleError : ArenaBindingError
{
    public ArenaState? State { get; }

    public ArenaLifecycleError(string message, ArenaState? state) : base(message)
    {
        State = state;
    }

    public static ArenaBindingError From(ArenaBindingError error)
    {
        if (error is ArenaLifecycleError)
            return error;
        var document = error.StateDocument;
        if (string.IsNullOrEmpty(document))
            return error;
        ArenaState state;
        try
        {
            state = ArenaState.Parse(document!);
        }
        catch (System.ArgumentException)
        {
            return error;
        }
        return new ArenaLifecycleError(error.Message, state);
    }
}

File: arena-junit/src/main/java/arena/junit/lifecycle/ArenaLifecycleError.java
package arena.junit.lifecycle;

import arena.junit.ffi.ArenaBindingError;

public class ArenaLifecycleError extends ArenaBindingError {
  private final transient ArenaState state;

  public ArenaLifecycleError(String message, ArenaState state) {
    super(message);
    this.state = state;
  }

  public ArenaState state() {
    return state;
  }

  public static ArenaBindingError from(ArenaBindingError error) {
    if (error instanceof ArenaLifecycleError) {
      return error;
    }
    String document = error.stateDocument();
    if (document == null || document.isEmpty()) {
      return error;
    }
    ArenaState state;
    try {
      state = ArenaState.parse(document);
    } catch (IllegalArgumentException e) {
      return error;
    }
    return new ArenaLifecycleError(error.getMessage(), state);
  }
}
